### ISSUE TITLE
Hard to use on ByteStrings. expects tag names to be the same type

### ISSUE THREAD

**Issue by seanhess** [fimad/scalpel#4]
Hi, thanks for your hard work! Very useful library. 

I'm having a hard time using it to scrape a lazy ByteString. I'm using `scrape` after parsing the tags with tag soup. 

So I have to make all my scraping functions be `Scraper Bytestring`, right? In order to get this to work with OverloadedStrings , I have to add the `:: ByteString` to the tag name. Shouldn't it be able work without that? 

```
scrapeTitle :: Scraper ByteString ByteString
scrapeTitle = text ("title" :: ByteString)
```

It seems like the type of the tag name is tied to the type of the string I'm parsing. Is there any way to have the tag names always be the same type, and just convert it when you need to compare it later?

**Comment by fimad** [comment]
Hmmm, yeah is inconvenient. The `StringLike` type class defines a `fromString` function, so we could make all the parameters be `String`s and convert them under the hood to whatever the desired result type is.

**Comment by fimad** [comment]
I tried several approaches to get this working, breaking out instances into separate modules, and only allowing Strings in the public API. Unfortunately neither fully solves the problem.

From what I can tell the primary issue is that for functions like `text`,

    text :: (StringLike str, Selectable str s) => s -> Scraper str str

when you turn on `-XOverloadedStrings` the first parameter s is constrained by `IsString s => s` and `Selectable str s` and GHC cannot derive a concrete type.

The closest thing to working would be c53edc3b. It would allow you to use `ByteString`s as the underlying tag type and `String`s for the API calls. You'd still have to turn off `-XOverloadedStrings`, but you would be able to write the following:

    scrapeTitle :: Scraper ByteString ByteString
    scrapeTitle = text "title"

Let me know If you have any ideas for how to get this working.

**Comment by seanhess** [comment]
Would it be possible to explicitly type text to `String` and then convert it to the correct selectable later?

    text :: (TagSoup.StringLike str) => String -> Scraper str str

We'll probably have to pull in someone more experienced than me to figure this out.

**Comment by fimad** [comment]
The problem with that is that it would prevent you from being able to use more complex selectors. For example you couldn't do `text Any` or `text ("div" // "a")`

**Comment by fimad** [comment]
I merged c53edc3 and released version 0.2.0. This makes all the APIs take Strings regardless of the tag type. Unfortunately for now I think you will have to either turn off `-XOverloadedStrings` or be resigned to explicitly typing your string literals :(

Closing this out for now since I don't think there is anything else that can be done.

**Comment by benjaminweb** [comment]
As I understand it HTML needs to be converted into Text. Is this due to TagSoup?
Related: TagSoup's limits the performance significantly in my use-case: https://github.com/ndmitchell/tagsoup/issues/83

What is simpler? To improve TagSoup or create a replacement? Which functionality does TagSoup give Scalpel actually?
